### PR TITLE
Prevent timing attack on CSRF, completing wonderful pr by @eutopian

### PR DESCRIPTION
**Pull request by jhartzler**
Attempting to complete the CSRF fix introduced in #123 by @eutopian 
It looks like that PR has been untouched since 2021 so opening a new one that is up to date with latest master and implements the renamed test that was suggested

### CODE REVIEW

**Comment by jhartzler** [comment]
@BobbyMcWho @nov do either of you have write access? 🙏 
New contributor just trying to fix a vuln, if you are around thank you so much in advance for your time!

**Comment by jhartzler** [comment]
Let me fix this build error

File: lib/omniauth/strategies/oauth2.rb [hunk 1]
```diff
@@ -83,8 +83,7 @@ def token_params
 
       def callback_phase # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity
         error = request.params["error_reason"] || request.params["error"]
-        if !options.provider_ignores_state && (request.params["state"].to_s.empty? || request.params["state"] != session.delete("omniauth.state"))
-          fail!(:csrf_detected, CallbackError.new(:csrf_detected, "CSRF detected"))
+      elsif !options.provider_ignores_state && (request.params["state"].to_s.empty? || !secure_compare(request.params["state"], session.delete("omniauth.state")))
```
Comment (BobbyMcWho) [inline lib/omniauth/strategies/oauth2.rb:86]:
```suggestion
      if !options.provider_ignores_state && (request.params["state"].to_s.empty? || !secure_compare(request.params["state"], session.delete("omniauth.state")))
```

Comment (jhartzler) [reply lib/omniauth/strategies/oauth2.rb]: Thanks for the review. Looks like I didn't copy from the initial PR properly

**Comment by jhartzler** [comment]
Running specs locally it looks like there might be some more work to reconcile this PR with master. I'm looking into it

**Comment by jhartzler** [comment]
Alright everything should be fixed now as far as I can tell. Specs pass locally.

**Comment by jhartzler** [comment]
@BobbyMcWho thank you for the earlier review and code suggestion
Not sure what your cadence/availability is but when you get a chance could I get a workflow kickoff and a re review?

**Comment by jhartzler** [comment]
I will buy someone a coffee if they review this PR lol

**Comment by BobbyMcWho** [comment]
No coffee needed, I don't mind merging, but the whole release process is the headache

**Review by BobbyMcWho — APPROVED** [review]

**Comment by jhartzler** [comment]
Haha thank you for the review + merge
I'm guessing there's nothing I can do to help the release process along?

**Comment by BobbyMcWho** [comment]
I just haven't released on this laptop yet so I gotta do some pre requisite installs, that's all

**Comment by BobbyMcWho** [comment]
Release [v1.9.0](https://github.com/omniauth/omniauth-oauth2/releases/tag/v1.9.0) available on [Rubygems](https://rubygems.org/gems/omniauth-oauth2/versions/1.9.0)

**Pull request merged**